File: cpp/n-ary-tree-postorder-traversal.cpp
/*
// Definition for a Node.
class Node {
public:
    int val;
    vector<Node*> children;

    Node() {}

    Node(int _val, vector<Node*> _children) {
        val = _val;
        children = _children;
    }
};
*/
// Recursively without DFS
class Solution {
public:
    vector<int> postorder(Node* root) {
        if (!root) return {};
        vector<int> res;
        vector<Node*> nodes = {root};
        while (!nodes.empty()) {
            Node* back = nodes.back();
            nodes.pop_back();
            res.push_back(back->val);
            nodes.insert(nodes.end(),
                         back->children.begin(),
                         back->children.end());
        }
        reverse(res.begin(), res.end());
        return res;
    }
};


// Recursively using DFS
class Solution {
public:
    void dfs(Node* root, vector<int>& ret) {
        if (!root) { return ; }

        for (auto& child : root->children) {
            dfs(child, ret);
        }
        ret.push_back(root->val);
    }
    vector<int> postorder(Node* root) {
        vector<int> ret;
        dfs(root, ret);
        return ret;
    }
};


// Iteratively
class Solution {
public:
    vector<int> postorder(Node* root) {
        stack<Node*> st;
        st.push(root);
        vector<int> result;
        while (!st.empty() and root) {
            Node* n = st.top();
            st.pop();
            result.insert(result.begin(), n->val);
            for (auto& child : n->children) {
                st.push(child);
            }
        }
        return result;
    }
};

File: py/n-ary-tree-postorder-traversal.py
"""
# Definition for a Node.
class Node:
    def __init__(self, val, children):
        self.val = val
        self.children = children
"""

# Recursively
class Solution:
    def postorder(self, root: 'Node') -> List[int]:
        if not root:
            return []
        if root.children is None:
            return [root.val]

        result = []
        res = []
        for child in root.children:
            res += self.postorder(child)
        result += res
        result.append(root.val)
        return result

# Iteratively
class Solution:
    def postorder(self, root: 'Node') -> List[int]:
        stack = [root]
        res = []
        while stack and root:
            n = stack.pop()
            res.insert(0, n.val)
            stack += n.children
        return res

# Recursively with post_order_traversal function
class Solution:
    def postorder(self, root: 'Node') -> List[int]:
        res = []
        def post_order_traversal(node: 'Node') -> None:
            if not node:
                return
            for child in node.children:
                post_order_traversal(child)
            res.append(node.val)
        post_order_traversal(root)
        return res


# Iteratively with helper function to minimize work
class Solution:
    def postorder(self, root: 'Node') -> List[int]:
        res = []

        def traverse_and_check(x: 'Node'):
            # lets us reference it outside the scope without
             # passing it as a parameter
            nonlocal res
            if x:
                # if x has no children, just append x's val
                if not x.children:
                    res.append(x.val)
                else: # else append every child for every child
                    for child in x.children:
                        traverse_and_check(child)
                    res.append(x.val)
        # End of traverse_and_check()
        if root:
            traverse_and_check(root)
        return res









